core/raftlease/fsm.go

```diff
@@ -158,11 +158,24 @@ func (f *FSM) GlobalTime() time.Time {
 	return f.globalTime
 }
 
-// Leases gets information about all of the leases in the system.
-func (f *FSM) Leases(localTime time.Time) map[lease.Key]lease.Info {
-	f.mu.Lock()
+// Leases gets information about all of the leases in the system,
+// optionally filtered by the input lease keys.
+func (f *FSM) Leases(localTime time.Time, keys ...lease.Key) map[lease.Key]lease.Info {
+	filter := make(map[lease.Key]bool)
+	filtering := len(keys) > 0
+	if filtering {
+		for _, key := range keys {
+			filter[key] = true
+		}
+	}
+
 	results := make(map[lease.Key]lease.Info)
+	f.mu.Lock()
 	for key, entry := range f.entries {
+		if filtering && !filter[key] {
```

Comment:
Oh, this just occurred to me - this would be far more efficient if it looped over the filter entries and gets them from f.entries (if they're there), especially given that we only specify one in the place it's used. It complicates the code but the savings are probably worth it for big controllers. Could you make that change please?

Reply:
Good catch. Will modify.